File: week3/Eren-Taskiran/25-reverse-nodes-in-k-group.java
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k <= 1) {
            return head;
        }
        ListNode current = head;
        int count = 0;
        while (current != null && count < k) {
            current = current.next;
            count++;
        }
        if (count < k) {
            return head;
        }
        ListNode reversedHead = reverseFirstK(head, k);
        head.next = reverseKGroup(current, k);
        return reversedHead;
    }
    
    public ListNode reverseFirstK(ListNode head, int k) {
        ListNode prev = null;
        ListNode current = head;
        int count = 0;
        while (current != null && count < k) {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
            count++;
        }
        return prev;
    }
}

File: week3/atakantekatan/25. Reverse Nodes in k-Group.py
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        first = dummy
        f = dummy
        rev = ListNode(0)
        second = rev
        while first.next:
            cnt = 0
            r = None
            while first.next and cnt < k:
                new = ListNode(first.next.val)
                if r is None:
                    r = new
                else:
                    new.next = r
                    r = new
                cnt += 1
                first = first.next
            if cnt == k:
                second.next = r
                for i in range(k):
                    f = f.next
            while second.next:
                second = second.next
        second.next = f.next
        return rev.next